packages/admin/cms-admin/src/dam/Table/useFolderTableQuery.ts

```diff
@@ -0,0 +1,97 @@
+import { QueryResult, useQuery } from "@apollo/client";
+import { IFilterApi } from "@comet/admin";
+
+import { GQLDamFileTableFragment, GQLDamFolderTableFragment, GQLDamItemsListQuery, GQLDamItemsListQueryVariables } from "../../graphql.generated";
+import { DamFilter } from "../DamTable";
+import { damItemsListQuery } from "./FolderTable.gql";
+
+export const damItemsListLimit = 100;
+
+const mergeDamItemsList = (previousResult: GQLDamItemsListQuery, newResult: GQLDamItemsListQuery, args: GQLDamItemsListQueryVariables) => {
+    const existing = previousResult.damItemsList;
+    const incoming = newResult.damItemsList;
+
+    let merged = existing ? { ...existing, nodes: [...existing.nodes] } : { nodes: [], totalCount: 0 };
+    if (args) {
+        const offset = args.offset ?? 0;
+        // eslint-disable-next-line @typescript-eslint/no-unused-vars
+        const { nodes, ...restProps } = incoming;
+        if (nodes != null) {
+            for (let i = 0; i < incoming.nodes.length; ++i) {
+                merged.nodes[offset + i] = incoming.nodes[i];
+            }
+
+            merged = { ...restProps, nodes: [...merged.nodes] };
+        }
+    } else {
+        merged = { ...existing, ...incoming, nodes: [...merged.nodes, ...incoming.nodes] };
+    }
+    return { damItemsList: merged };
+};
+
+interface FolderTableQueryProps {
+    folderId?: string;
+    filterApi: IFilterApi<DamFilter>;
+    allowedMimetypes?: string[];
+}
+
+export interface FolderTableData {
+    data: Array<GQLDamFileTableFragment | GQLDamFolderTableFragment>;
+    totalCount: number;
+}
+
+export interface FolderTableQueryApi extends QueryResult<GQLDamItemsListQuery, GQLDamItemsListQueryVariables> {
+    tableData?: FolderTableData;
+}
+
+export const useFolderTableQuery = ({ folderId, filterApi, allowedMimetypes }: FolderTableQueryProps): FolderTableQueryApi => {
+    const { data, fetchMore, ...rest } = useQuery<GQLDamItemsListQuery, GQLDamItemsListQueryVariables>(damItemsListQuery, {
+        variables: {
+            folderId: folderId,
+            includeArchived: filterApi.current.archived,
+            filter: {
+                mimetypes: allowedMimetypes,
+                searchText: filterApi.current.searchText,
+            },
+            sortColumnName: filterApi.current.sort?.columnName,
+            sortDirection: filterApi.current.sort?.direction,
+            limit: damItemsListLimit,
+            offset: 0,
+        },
+        onCompleted: async (data) => {
+            if (data) {
+                const currentDataLength = data.damItemsList.nodes.length;
+                const totalCount = data.damItemsList.totalCount;
+
+                if (totalCount > currentDataLength) {
+                    await fetchMore<GQLDamItemsListQuery, GQLDamItemsListQueryVariables>({
```

Comment:
what are you doing here, right in onCompleted you call fetchMore?

Reply:
Ok, I think I understand: you want to avoid loading all entries in one request, instead you split it up into chunks.

Why not use a standard paging? This technique will also have it's limitations, and still causes the same server load, just over a longer time period (actually even more, as it has more overhead for more requests)

I mean, we should also support 10.000 enties

Reply:
> you want to avoid loading all entries in one request, instead you split it up into chunks

correct

> This technique will also have it's limitations, and still causes the same server load, just over a longer time period (actually even more, as it has more overhead for more requests)

I implemented it because @johnnyomair and @chernylu reported that the API crashed when trying to load 4000 items at once. 

Overall, this technique will cause the same load. But the load is spread over a longer period of time in a controlled way.

However, this technique will, of course, be slow if there are a lot of items (especially the refetches will be annoying).

> Why not use a standard paging?

I suppose you mean paging with actual pages and back/forward buttons (?)

This would also cause some problems:

- **Drag and Drop:** Currently folders are displayed first. Files are shown after the folders. The only way to move a file into a folder is via Drag and Drop. 
When using standard paging, it might not be possible to move a file into a folder. E.g. if all the folders are on page 1 and but the file is on page 3.

- **Locating new files:** Currently, we have the functionality that after adding a new file in the current folder, this file is automatically scrolled into view and highlighted. 
When using standard paging, this is harder to implement as the file has to be located on its page.

- **Multiselect (Select all):** We have this select all checkbox at the top of the DAM. When clicking it, the expected behavior is that all items in the current folder are marked. However, when using real paging, we don't have all the items on the client-side. 
Therefore, multiselect would have to be implemented in the API.
There are also some UX challenges, e.g., you first click select all. Then you unselect one item. Are now all items in the folder (minus that item) selected or all items on the current page (minus that item)?

- **UX:** Nobody expects to have pages in a file system. 
Our file system is already divided into folders. Entering a folder and then navigating multiple pages inside this folder seems a bit tedious. 
In my opinion, it's better to have a slow loading process and maybe recommend the user to split the files into folders if it's to slow via a hint shown while loading.

> I mean, we should also support 10.000 enties

I'm also not completely convinced, that the solution introduced in this PR is the perfect solution forever.
@johnnyomair and @chernylu approached me and told me that we need a workaround for a client project by Wednesday, 28.09.
Therefore, I looked for a solution that worked and could be implemented quickly.

We should take another look at this in the future.
But a more permanent solution (like real infinite scrolling (?)) will take some more development time because a lot of existing functionality has to be adjusted.

Reply:
Yes, everything is easier without paging :)

But it simply doesn't scale. Probably some features have to be re-thought with paging in mind.

Reply:
I think we should discuss this with UX. I've created a task for it